internal/venice-common/src/main/java/com/linkedin/venice/writer/CompositeVeniceWriter.java

```diff
@@ -0,0 +1,182 @@
+      boolean hasPutMetadata) {
+    CompletableFuture<PubSubProduceResult> finalFuture = new CompletableFuture<>();
+    CompletableFuture[] childFutures = new CompletableFuture[childWriters.length + 1];
+    int index = 0;
+    childFutures[index++] = lastWriteFuture;
+    for (VeniceWriter<K, V, U> writer: childWriters) {
+      childFutures[index++] = hasPutMetadata
+          ? writer.put(
+              key,
+              value,
+              valueSchemaId,
+              childCallback,
+              DEFAULT_LEADER_METADATA_WRAPPER,
+              APP_DEFAULT_LOGICAL_TS,
+              putMetadata)
+          : writer.put(key, value, valueSchemaId, childCallback);
+    }
+    CompletableFuture.allOf(childFutures).whenCompleteAsync((ignored, childException) -> {
+      if (childException == null) {
+        CompletableFuture<PubSubProduceResult> mainFuture = hasPutMetadata
+            ? mainWriter.put(
+                key,
+                value,
+                valueSchemaId,
+                callback,
+                DEFAULT_LEADER_METADATA_WRAPPER,
+                APP_DEFAULT_LOGICAL_TS,
+                putMetadata)
+            : mainWriter.put(key, value, valueSchemaId, callback);
+        mainFuture.whenCompleteAsync((result, mainWriteException) -> {
+          if (mainWriteException == null) {
+            finalFuture.complete(result);
+          } else {
+            finalFuture.completeExceptionally(new VeniceException(mainWriteException));
+          }
+        });
+      } else {
+        VeniceException veniceException = new VeniceException(childException);
+        finalFuture.completeExceptionally(veniceException);
+      }
+    });
+    lastWriteFuture = finalFuture;
+    return finalFuture;
+  }
+
+  @Override
+  public CompletableFuture<PubSubProduceResult> delete(
+      K key,
+      PubSubProducerCallback callback,
+      DeleteMetadata deleteMetadata) {
+    CompletableFuture<PubSubProduceResult> finalFuture = new CompletableFuture<>();
+    CompletableFuture[] childFutures = new CompletableFuture[childWriters.length + 1];
+    int index = 0;
+    childFutures[index++] = lastWriteFuture;
+    for (VeniceWriter<K, V, U> writer: childWriters) {
+      childFutures[index++] = writer.delete(key, callback, deleteMetadata);
+    }
+    CompletableFuture.allOf(childFutures).whenCompleteAsync((ignored, childException) -> {
+      if (childException == null) {
+        CompletableFuture<PubSubProduceResult> mainFuture = mainWriter.delete(key, callback, deleteMetadata);
+        mainFuture.whenCompleteAsync((result, mainWriteException) -> {
+          if (mainWriteException == null) {
+            finalFuture.complete(result);
+          } else {
+            finalFuture.completeExceptionally(new VeniceException(mainWriteException));
+          }
+        });
+      } else {
+        VeniceException veniceException = new VeniceException(childException);
+        finalFuture.completeExceptionally(veniceException);
```

Comment:
I can see some duplicate code in functions: `put` and `delete`, so is it easy to de-dup them?

Reply:
Yes, much cleaner now after I refactored the methods using `java.util.function.Function`.